kaitaistruct.py

```diff
@@ -299,28 +299,42 @@ def read_bytes(self, n):
                 "requested invalid %d amount of bytes" %
                 (n,)
             )
-
-        is_satisfiable = True
-        # in Python 2, there is a common error ['file' object has no
-        # attribute 'seekable'], so we need to make sure that seekable() exists
-        if callable(getattr(self._io, 'seekable', None)) and self._io.seekable():
-            num_bytes_available = self.size() - self.pos()
-            is_satisfiable = (n <= num_bytes_available)
-
-        if is_satisfiable:
+        elif n < 8*1024*1024:  # = 8 MiB
+            # For relatively small amounts of data, it's safe and faster
+            # to read the data unconditionally and check its size afterwards.
             r = self._io.read(n)
-            num_bytes_available = len(r)
-            is_satisfiable = (n <= num_bytes_available)
+            if len(r) < n:
+                raise EOFError(
+                    "requested %d bytes, but got only %d bytes" %
+                    (n, len(r))
+                )
+            return r
+        else:
+            # When a large number of bytes is requested, try to check first
+            # that there is indeed enough data left in the stream to satisfy
+            # the read. This avoids reading large amounts of data only to notice
+            # afterwards that it's not long enough.
+            is_satisfiable = True
+            # in Python 2, there is a common error ['file' object has no
+            # attribute 'seekable'], so we need to make sure that seekable() exists
+            if callable(getattr(self._io, 'seekable', None)) and self._io.seekable():
```

Comment:
You're completely right. I took the before/after test that I did and translated it very literally into an `if`, without thinking about how it could be simplified 🙂